### DISCUSSION TITLE
Should this be Labeled as ERROR or Just an WARN / NOTIFICATION?

### DISCUSSION THREAD

**Discussion by Ashu999** [prisma/prisma#17185] [Q&A] [answered]
I'm running a node.js server+prisma in Docker, This ERROR: `The migration '0_init' is already recorded as applied in the database` Is causing the container to stop. I Think it should not be an ERROR but just a WARN / NOTIFICATION.
It's behavior should be similar to `npx prisma migrate deploy`, which just tells us `No pending migrations to apply.`, rather than throwing an ERROR.
Just wanted to discuss, why it was decided to show it as an ERROR. Thanks

```
server_1       | > npx prisma migrate resolve --applied 0_init && npx prisma migrate deploy && node dist/index.js
server_1       | 
server_1       | Environment variables loaded from prisma/.env
server_1       | Prisma schema loaded from prisma/schema.prisma
server_1       | Datasource "db": MySQL database "supertokens" at "db:3306"
server_1       | Error: P3008
server_1       | 
server_1       | The migration `0_init` is already recorded as applied in the database.
server_1 exited with code 1
```

**Comment by janpio** [comment, marked as answer]
You tell Prisma to mark the `0_init` migration as resolved "applied" in the migration history. This implies that it is not marked as applied yet. If it is though, this indicates that your intent of running the command can not be met.

Why do you _always_ want to run this command?

**Reply by Ashu999** [reply]
I don't _always_ want to run this command, but only once (such that mysql has started). but I could find no other way. Let me explain:
My docker-compose has mysql, server in it:
```
-mysql
-server
   depends_on: mysql
   restart: unless-stopped
```
The prima migrations commands have to be run when mysql has started. So I put those commands in `CMD` of Dockerfile.server.
Sometimes server is not able to find mysql, because it takes time to start (even though it's container is running). So I have put `restart: unless-stopped `in the server, to keep re-trying. 
the CMD basically runs this:
`npx prisma migrate resolve --applied 0_init && npx prisma migrate deploy && node dist/index.js`
After running the containers once, mysql persist the `npx prisma migrate resolve --applied 0_init` change. 
So next time running the containers, I have to modify the cmd to `npx prisma migrate deploy && node dist/index.js`, and rebuild. Otherwise, I'll get the error `The migration '0_init' is already recorded as applied in the database`.
I want to avoid modifying the `CMD`. which would have been possible if `npx prisma migrate resolve --applied 0_init` didn't throw an error.

Let me know if you didn't understand what I'm saying. Or if there is a better way to do it. I have not worked with Prisma and docker before.
Thanks.

**Reply by janpio** [reply]
I am roughly following.

Why do you need to baseline and use `resolve --applied` when you start your MySQL as a docker container? Does it already have existing schema and data in there?

**Reply by Ashu999** [reply]
> Does it already have existing schema and data in there?

yes.
Why?
I'm using a user authentication service ([supertokens](https://supertokens.com/docs/thirdparty/custom-ui/init/database-setup/mysql)). The schema is created by this service, This service is also part of docker-compose.
Maybe I can just revoke the table creation permission from supertokens, and do it through prima only. That will work

**Reply by janpio** [reply]
I think I understood your original description now, it makes sense to me how you approached this.

> Why?

I am asking to understand, that is not the most common use case that there already is something in the database, and then you need to run baselining. 

What exactly is in `0_init/migration.sql` - what the authentication service already created only or more than that?

**Reply by Ashu999** [reply]
`0_init/migration.sql` is what the authentication service already created only, nothing more than that.
I used `npx prisma db pull` . As mentioned [here](https://www.prisma.io/docs/getting-started/setup-prisma/add-to-existing-project/relational-databases/introspection-typescript-mysql).

btw when I said 'Why', I was not asking. I was saying it like: "why? _you may ask_".

**Reply by janpio** [reply]
I think I now fully understand your problem and use case, and will try to reflect it in an issue. 
That will make it easier for us to talk about this internally to see if there is a solution.

**Reply by janpio** [reply]
Short term, you can probably adapt your command you run on startup:
```
npx prisma migrate resolve --applied 0_init && npx prisma migrate deploy && node dist/index.js
```
to
```
npx prisma migrate resolve --applied 0_init; npx prisma migrate deploy && node dist/index.js
```
Source: https://www.geeksforgeeks.org/difference-between-chaining-operators-in-linux/ (Yep, I googled that as a Windows user - to please try if it actually works)

**Reply by janpio** [reply]
Actually, what happens when you run `npx prisma db pull` when you have all migrations, both the one from supertoken and the ones from Prisma applied? Does this create a Prisma model that represents all the information from the database that is required to use supertoken? Or does the database image actually do _more_ than just  create a few models?

**Reply by Ashu999** [reply]
> Actually, what happens when you run npx prisma db pull when you have all migrations, both the one from supertoken and the ones from Prisma applied? 

when I do `npx prisma db pull`, I'm only pulling the schema change done by supertokens, and creating a migration for it `0_init `. When I add new models, I create new migration on top of it using `npx prisma migrate dev`.  `0_init ` is the base/first migration.
So, my CMD to apply these changes on the docker container looks like this: `npx prisma migrate resolve --applied 0_init && npx prisma migrate deploy && node dist/index.js`

> Does this create a Prisma model that represents all the information from the database that is required to use supertoken?

Yes, This creates a Prisma model that represents all the information from the database that is required to use supertoken. I have also tested this, I had once reset/overwritten the MySQL DB using changes that were pulled into `prisma.schema`. Everything worked fine.

> Or does the database image actually do more than just create a few models?

It does more than just create the schema. It also populates data in at least one of the table `key_value` (as mentioned in docs [here](https://supertokens.com/docs/thirdparty/custom-ui/init/database-setup/mysql)).  But seems like this table gets populated again by the supertokens service. Basically, they have also handled the case where DB schema is not created by their service.

**Reply by Ashu999** [reply]
> I think I now fully understand your problem and use case, and will try to reflect it in an issue. That will make it easier for us to talk about this internally to see if there is a solution.

Let me know If I can contribute. Thanks.

**Reply by janpio** [reply]
Did the suggested command work for you?

**Reply by janpio** [reply]
I created an issue at https://github.com/prisma/prisma/issues/17314, although if this command worked, I might be saying that is a valid workaround for now.

**Reply by Ashu999** [reply]
Yes, this command works, a valid workaround: https://www.geeksforgeeks.org/difference-between-chaining-operators-in-linux/
Thanks.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
